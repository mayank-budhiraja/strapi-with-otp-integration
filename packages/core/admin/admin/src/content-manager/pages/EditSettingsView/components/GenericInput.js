import React from 'react';
import PropTypes from 'prop-types';
import { TextInput } from '@strapi/design-system/TextInput';
import { ToggleInput } from '@strapi/design-system/ToggleInput';
import { Select, Option } from '@strapi/design-system/Select';
import { useIntl } from 'react-intl';

const GenericInput = ({ type, options, onChange, value, name, ...inputProps }) => {
  const { formatMessage } = useIntl();

  switch (type) {
    case 'text': {
      return <TextInput onChange={onChange} value={value} name={name} {...inputProps} />;
    }
    case 'bool': {
      return (
        <ToggleInput
          onChange={e => {
            onChange({ target: { name, value: e.target.checked } });
          }}
          checked={value}
          name={name}
          onLabel={formatMessage({
            id: 'app.components.ToggleCheckbox.on-label',
            defaultMessage: 'On',
          })}
          offLabel={formatMessage({
            id: 'app.components.ToggleCheckbox.off-label',
            defaultMessage: 'Off',
          })}
          {...inputProps}
        />
      );
    }
    case 'select': {
      return (
        <Select
          value={value}
          name={name}
          onChange={value => onChange({ target: { name, value } })}
          {...inputProps}
        >
          {options.map(option => (
            <Option key={option} value={option}>
              {option}
            </Option>
          ))}
        </Select>
      );
    }
    default:
      return null;
  }
};

GenericInput.defaultProps = {
  options: undefined,
};

GenericInput.propTypes = {
  type: PropTypes.string.isRequired,
  options: PropTypes.arrayOf(PropTypes.string),
  onChange: PropTypes.func.isRequired,
  value: PropTypes.oneOfType([PropTypes.string, PropTypes.bool]).isRequired,
  name: PropTypes.string.isRequired,
};

export default GenericInput;
