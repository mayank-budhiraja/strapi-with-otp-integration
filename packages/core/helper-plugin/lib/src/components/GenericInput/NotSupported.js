/**
 *
 * NotSupported
 *
 */

import React from 'react';
import { TextInput } from '@strapi/design-system/TextInput';
import PropTypes from 'prop-types';

const NotSupported = ({ hint, label, labelAction, error, name, required }) => {
  return (
    <TextInput
      disabled
      error={error}
      label={label}
      labelAction={labelAction}
      id={name}
      hint={hint}
      name={name}
      onChange={() => {}}
      placeholder="Not supported"
      required={required}
      type="text"
      value=""
    />
  );
};

NotSupported.defaultProps = {
  hint: null,
  error: undefined,
  labelAction: undefined,
  required: false,
};

NotSupported.propTypes = {
  error: PropTypes.string,
  hint: PropTypes.string,
  label: PropTypes.string.isRequired,
  labelAction: PropTypes.element,
  name: PropTypes.string.isRequired,
  required: PropTypes.bool,
};

export default NotSupported;
