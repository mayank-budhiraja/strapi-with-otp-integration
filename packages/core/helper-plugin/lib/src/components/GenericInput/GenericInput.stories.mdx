<!--- GenericInput.stories.mdx --->

import { useState } from 'react';
import { Meta, Story, Canvas } from '@storybook/addon-docs/blocks';
import { Grid, GridItem } from '@strapi/design-system/Grid';
import { HeaderLayout, ContentLayout } from '@strapi/design-system/Layout';
import { Main } from '@strapi/design-system/Main';
import { Stack } from '@strapi/design-system/Stack';
import { Box } from '@strapi/design-system/Box';
import set from 'lodash/set';
import GenericInput from './index';
import layout from './storyUtils/layout';

<Meta title="components/GenericInput" component={GenericInput} />

# GenericInput

This is the doc of the `GenericInput` component

## GenericInput

Description...

<Canvas>
  <Story name="base">
    {() => {
      const [state, setState] = useState({
        firstname: 'ee',
        email: '',
        private: null,
        isNullable: null,
        isActiveByDefault: true,
        isInactiveByDefault: false
      });
      return (
        <ContentLayout>
          <Stack spacing={4}>
            {layout.map((row, rowIndex) => {
              return (
                <Box key={rowIndex}>
                  <Grid gap={4}>
                    {row.map(input => {
                      return (
                        <GridItem key={input.name} {...input.size}>
                          <GenericInput
                            {...input}
                            value={state[input.name]}
                            onChange={({ target: { name, value } }) => {
                              setState(prev => {
                                return { ...prev, [name]: value };
                              });
                            }}
                          />
                        </GridItem>
                      );
                    })}
                  </Grid>
                </Box>
              );
            })}
          </Stack>
        </ContentLayout>
      );
    }}
  </Story>
</Canvas>
