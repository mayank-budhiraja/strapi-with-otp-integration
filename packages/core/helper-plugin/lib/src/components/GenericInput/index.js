/**
 *
 * GenericInput
 *
 */

import React, { useState } from 'react';
import PropTypes from 'prop-types';
import parseISO from 'date-fns/parseISO';
import formatISO from 'date-fns/formatISO';
import { useIntl } from 'react-intl';
import { Checkbox } from '@strapi/design-system/Checkbox';
import { DatePicker } from '@strapi/design-system/DatePicker';
import { NumberInput } from '@strapi/design-system/NumberInput';
import { Select, Option } from '@strapi/design-system/Select';
import { Textarea } from '@strapi/design-system/Textarea';
import { TextInput } from '@strapi/design-system/TextInput';
import { TimePicker } from '@strapi/design-system/TimePicker';
import { ToggleInput } from '@strapi/design-system/ToggleInput';
import { Icon } from '@strapi/design-system/Icon';
import EyeStriked from '@strapi/icons/EyeStriked';
import Eye from '@strapi/icons/Eye';
import DateTimePicker from '../DateTimePicker';
import NotSupported from './NotSupported';

const GenericInput = ({
  autoComplete,
  customInputs,
  description,
  disabled,
  intlLabel,
  labelAction,
  error,
  name,
  onChange,
  options,
  placeholder,
  required,
  step,
  type,
  value: defaultValue,
  isNullable,
  ...rest
}) => {
  const { formatMessage } = useIntl();
  const [showPassword, setShowPassword] = useState(false);

  const CustomInput = customInputs ? customInputs[type] : null;

  // the API always returns null, which throws an error in React,
  // therefore we cast this case to undefined
  const value = defaultValue ?? undefined;

  /*
   TODO: ideally we should pass in `defaultValue` and `value` for
   inputs, in order to make them controlled components. This variable
   acts as a fallback for now, to prevent React errors in devopment mode

   See: https://github.com/strapi/strapi/pull/12861
  */
  const valueWithEmptyStringFallback = value ?? '';

  function getErrorMessage(error) {
    if (!error) {
      return null;
    }

    const values = {
      ...error.values,
    };

    if (typeof error === 'string') {
      return formatMessage({ id: error, defaultMessage: error }, values);
    }

    return formatMessage(
      {
        id: error.id,
        defaultMessage: error?.defaultMessage ?? error.id,
      },
      values
    );
  }

  const errorMessage = getErrorMessage(error);

  if (CustomInput) {
    return (
      <CustomInput
        {...rest}
        description={description}
        disabled={disabled}
        intlLabel={intlLabel}
        labelAction={labelAction}
        error={errorMessage}
        name={name}
        onChange={onChange}
        options={options}
        required={required}
        placeholder={placeholder}
        type={type}
        value={value}
      />
    );
  }

  const label = intlLabel.id
    ? formatMessage(
        { id: intlLabel.id, defaultMessage: intlLabel.defaultMessage },
        { ...intlLabel.values }
      )
    : name;

  const hint = description
    ? formatMessage(
        { id: description.id, defaultMessage: description.defaultMessage },
        { ...description.values }
      )
    : '';

  const formattedPlaceholder = placeholder
    ? formatMessage(
        { id: placeholder.id, defaultMessage: placeholder.defaultMessage },
        { ...placeholder.values }
      )
    : '';

  switch (type) {
    case 'bool': {
      const clearProps = {
        clearLabel:
          isNullable &&
          formatMessage({
            id: 'app.components.ToggleCheckbox.clear-label',
            defaultMessage: 'Clear',
          }),

        onClear:
          isNullable &&
          (() => {
            onChange({ target: { name, value: null } });
          }),
      };

      return (
        <ToggleInput
          checked={defaultValue === null ? null : defaultValue || false}
          disabled={disabled}
          hint={hint}
          label={label}
          error={errorMessage}
          labelAction={labelAction}
          name={name}
          offLabel={formatMessage({
            id: 'app.components.ToggleCheckbox.off-label',
            defaultMessage: 'False',
          })}
          onLabel={formatMessage({
            id: 'app.components.ToggleCheckbox.on-label',
            defaultMessage: 'True',
          })}
          onChange={e => {
            onChange({ target: { name, value: e.target.checked } });
          }}
          required={required}
          {...clearProps}
        />
      );
    }
    case 'checkbox': {
      return (
        <Checkbox
          disabled={disabled}
          error={errorMessage}
          hint={hint}
          id={name}
          name={name}
          onValueChange={value => {
            onChange({ target: { name, value } });
          }}
          required={required}
          value={Boolean(value)}
        >
          {label}
        </Checkbox>
      );
    }
    case 'datetime': {
      return (
        <DateTimePicker
          clearLabel={formatMessage({ id: 'clearLabel', defaultMessage: 'Clear' })}
          disabled={disabled}
          error={errorMessage}
          label={label}
          labelAction={labelAction}
          id={name}
          hint={hint}
          name={name}
          onChange={date => {
            const formattedDate = date.toISOString();

            onChange({ target: { name, value: formattedDate, type } });
          }}
          step={step}
          onClear={() => onChange({ target: { name, value: null, type } })}
          placeholder={formattedPlaceholder}
          required={required}
          value={value && new Date(value)}
          selectedDateLabel={formattedDate => `Date picker, current is ${formattedDate}`}
        />
      );
    }
    case 'date': {
      let selectedDate = null;

      if (value) {
        selectedDate = parseISO(value);
      }

      return (
        <DatePicker
          clearLabel={formatMessage({ id: 'clearLabel', defaultMessage: 'Clear' })}
          disabled={disabled}
          error={errorMessage}
          label={label}
          labelAction={labelAction}
          id={name}
          hint={hint}
          name={name}
          onChange={date => {
            onChange({
              target: { name, value: formatISO(date, { representation: 'date' }), type },
            });
          }}
          onClear={() => onChange({ target: { name, value: null, type } })}
          placeholder={formattedPlaceholder}
          required={required}
          selectedDate={selectedDate}
          selectedDateLabel={formattedDate => `Date picker, current is ${formattedDate}`}
        />
      );
    }
    case 'number': {
      return (
        <NumberInput
          disabled={disabled}
          error={errorMessage}
          label={label}
          labelAction={labelAction}
          id={name}
          hint={hint}
          name={name}
          onValueChange={value => {
            onChange({ target: { name, value: value ?? null, type } });
          }}
          placeholder={formattedPlaceholder}
          required={required}
          step={step}
          value={value}
        />
      );
    }
    case 'email': {
      return (
        <TextInput
          autoComplete={autoComplete}
          disabled={disabled}
          error={errorMessage}
          label={label}
          labelAction={labelAction}
          id={name}
          hint={hint}
          name={name}
          onChange={onChange}
          placeholder={formattedPlaceholder}
          required={required}
          type="email"
          value={valueWithEmptyStringFallback}
        />
      );
    }
    case 'timestamp':
    case 'text':
    case 'string': {
      return (
        <TextInput
          autoComplete={autoComplete}
          disabled={disabled}
          error={errorMessage}
          label={label}
          labelAction={labelAction}
          id={name}
          hint={hint}
          name={name}
          onChange={onChange}
          placeholder={formattedPlaceholder}
          required={required}
          type="text"
          value={valueWithEmptyStringFallback}
        />
      );
    }
    case 'password': {
      return (
        <TextInput
          autoComplete={autoComplete}
          disabled={disabled}
          error={errorMessage}
          endAction={
            <button
              aria-label={formatMessage({
                id: 'Auth.form.password.show-password',
                defaultMessage: 'Show password',
              })}
              onClick={() => {
                setShowPassword(prev => !prev);
              }}
              style={{
                border: 'none',
                padding: 0,
                background: 'transparent',
              }}
              type="button"
            >
              {showPassword ? (
                <Icon as={Eye} color="neutral500" />
              ) : (
                <Icon as={EyeStriked} color="neutral500" />
              )}
            </button>
          }
          label={label}
          labelAction={labelAction}
          id={name}
          hint={hint}
          name={name}
          onChange={onChange}
          placeholder={formattedPlaceholder}
          required={required}
          type={showPassword ? 'text' : 'password'}
          value={valueWithEmptyStringFallback}
        />
      );
    }
    case 'select': {
      return (
        <Select
          disabled={disabled}
          error={errorMessage}
          label={label}
          labelAction={labelAction}
          id={name}
          hint={hint}
          name={name}
          onChange={value => {
            onChange({ target: { name, value: value === '' ? null : value, type: 'select' } });
          }}
          placeholder={formattedPlaceholder}
          required={required}
          value={value}
        >
          {options.map(({ metadatas: { intlLabel, disabled, hidden }, key, value }) => {
            return (
              <Option key={key} value={value} disabled={disabled} hidden={hidden}>
                {formatMessage(intlLabel)}
              </Option>
            );
          })}
        </Select>
      );
    }
    case 'textarea': {
      return (
        <Textarea
          disabled={disabled}
          error={errorMessage}
          label={label}
          labelAction={labelAction}
          id={name}
          hint={hint}
          name={name}
          onChange={onChange}
          required={required}
          placeholder={formattedPlaceholder}
          type={type}
          value={valueWithEmptyStringFallback}
        >
          {value}
        </Textarea>
      );
    }
    case 'time': {
      let time = value;

      // The backend send a value which has the following format: '00:45:00.000'
      // or the time picker only supports hours & minutes so we need to mutate the value
      if (value && value.split(':').length > 2) {
        time = time.split(':');
        time.pop();
        time = time.join(':');
      }

      return (
        <TimePicker
          clearLabel={formatMessage({ id: 'clearLabel', defaultMessage: 'Clear' })}
          disabled={disabled}
          error={errorMessage}
          label={label}
          labelAction={labelAction}
          id={name}
          hint={hint}
          name={name}
          onChange={time => {
            onChange({ target: { name, value: `${time}`, type } });
          }}
          onClear={() => {
            onChange({ target: { name, value: null, type } });
          }}
          placeholder={formattedPlaceholder}
          required={required}
          step={step}
          value={time}
        />
      );
    }
    default: {
      return (
        <NotSupported
          name={name}
          label={label}
          labelAction={labelAction}
          hint={hint}
          error={errorMessage}
          required={required}
        />
      );
    }
  }
};

GenericInput.defaultProps = {
  autoComplete: undefined,
  customInputs: null,
  description: null,
  disabled: false,
  error: '',
  isNullable: undefined,
  labelAction: undefined,
  placeholder: null,
  required: false,
  options: [],
  step: 1,
  value: undefined,
};

GenericInput.propTypes = {
  autoComplete: PropTypes.string,
  customInputs: PropTypes.object,
  description: PropTypes.shape({
    id: PropTypes.string.isRequired,
    defaultMessage: PropTypes.string.isRequired,
    values: PropTypes.object,
  }),
  disabled: PropTypes.bool,
  error: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.shape({
      id: PropTypes.string.isRequired,
      defaultMessage: PropTypes.string,
    }),
  ]),
  intlLabel: PropTypes.shape({
    id: PropTypes.string.isRequired,
    defaultMessage: PropTypes.string.isRequired,
    values: PropTypes.object,
  }).isRequired,
  isNullable: PropTypes.bool,
  labelAction: PropTypes.element,
  name: PropTypes.string.isRequired,
  onChange: PropTypes.func.isRequired,
  options: PropTypes.arrayOf(
    PropTypes.shape({
      metadatas: PropTypes.shape({
        intlLabel: PropTypes.shape({
          id: PropTypes.string.isRequired,
          defaultMessage: PropTypes.string.isRequired,
        }).isRequired,
        disabled: PropTypes.bool,
        hidden: PropTypes.bool,
      }).isRequired,
      key: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
      value: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
    }).isRequired
  ),
  placeholder: PropTypes.shape({
    id: PropTypes.string.isRequired,
    defaultMessage: PropTypes.string.isRequired,
    values: PropTypes.object,
  }),
  required: PropTypes.bool,
  step: PropTypes.number,
  type: PropTypes.string.isRequired,
  value: PropTypes.any,
};

export default GenericInput;
