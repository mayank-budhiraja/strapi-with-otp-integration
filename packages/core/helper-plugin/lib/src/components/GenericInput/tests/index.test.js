import React from 'react';
import { ThemeProvider, lightTheme } from '@strapi/design-system';
import { IntlProvider } from 'react-intl';
import { render, fireEvent } from '@testing-library/react';

import GenericInput from '../index';

function ComponentFixture(props) {
  return (
    <IntlProvider locale="en" messages={{}}>
      <ThemeProvider theme={lightTheme}>
        <GenericInput {...props} />
      </ThemeProvider>
    </IntlProvider>
  );
}

function setupNumber(props) {
  const NUMBER_FIXTURE_PROPS = {
    type: 'number',
    name: 'number',
    intlLabel: {
      id: 'label.test',
      defaultMessage: 'Default label',
    },
    placeholder: {
      id: 'placeholder.test',
      defaultMessage: 'Default placeholder',
    },
    hint: 'Hint message',
    value: null,
    required: true,
    onChange: jest.fn,
    ...props,
  };

  const rendered = render(<ComponentFixture {...NUMBER_FIXTURE_PROPS} />);
  const input = rendered.container.querySelector('input');

  return {
    ...rendered,
    input,
  };
}

describe('GenericInput', () => {
  describe('number', () => {
    test('renders and matches the snapshot', () => {
      const { container } = setupNumber();
      expect(container).toMatchSnapshot();
    });

    test('renders an error message', () => {
      const { getByText } = setupNumber({ error: 'Error message' });
      expect(getByText('Error message')).toBeInTheDocument();
    });

    test('renders a number (int) value', () => {
      const { container } = setupNumber({ value: 1 });
      expect(container.querySelector('[type="text"]').value).toBe('1');
    });

    test('renders a number (float) value', () => {
      const { container } = setupNumber({ value: 1.3333 });
      expect(container.querySelector('input').value).toBe('1.3333');
    });

    test('does not call onChange callback on first render', () => {
      const spy = jest.fn();
      setupNumber({ value: null, onChange: spy });
      expect(spy).not.toHaveBeenCalled();
    });

    test('does not call onChange callback if the value does not change', () => {
      const spy = jest.fn();
      const { input } = setupNumber({ value: 23, onChange: spy });

      fireEvent.change(input, { target: { value: 23 } });

      expect(spy).not.toHaveBeenCalledWith();
    });

    test('does call onChange callback with number (int) value', () => {
      const spy = jest.fn();
      const { input } = setupNumber({ value: null, onChange: spy });

      fireEvent.change(input, { target: { value: '23' } });

      expect(spy).toHaveBeenCalledWith({ target: { name: 'number', type: 'number', value: 23 } });
    });

    test('does call onChange callback with number (float) value', () => {
      const spy = jest.fn();
      const { input } = setupNumber({ value: null, onChange: spy });

      fireEvent.change(input, { target: { value: '1.3333' } });

      expect(spy).toHaveBeenCalledWith({
        target: { name: 'number', type: 'number', value: 1.3333 },
      });
    });

    test('does call onChange callback with number (0) value', () => {
      const spy = jest.fn();
      const { input } = setupNumber({ value: null, onChange: spy });

      fireEvent.change(input, { target: { value: '0' } });

      expect(spy).toHaveBeenCalledWith({
        target: { name: 'number', type: 'number', value: 0 },
      });
    });

    test('does call onChange callback with null on empty value', () => {
      const spy = jest.fn();
      const { input } = setupNumber({ value: 1, onChange: spy });

      fireEvent.change(input, { target: { value: '' } });

      expect(spy).toHaveBeenCalledWith({ target: { name: 'number', type: 'number', value: null } });
    });
  });
});
