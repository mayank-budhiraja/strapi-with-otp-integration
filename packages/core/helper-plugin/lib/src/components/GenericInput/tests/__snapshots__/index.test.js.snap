// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GenericInput number renders and matches the snapshot 1`] = `
.c14 {
  border: 0;
  -webkit-clip: rect(0 0 0 0);
  clip: rect(0 0 0 0);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  width: 1px;
}

.c9 {
  padding-right: 12px;
  padding-left: 8px;
}

.c11 {
  color: #8e8ea9;
}

.c0 {
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c3 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
}

.c6 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c2 {
  font-weight: 600;
  color: #32324d;
  font-size: 0.75rem;
  line-height: 1.33;
}

.c4 {
  color: #d02b20;
  font-size: 0.875rem;
  line-height: 1.43;
}

.c5 {
  line-height: 0;
}

.c8 {
  border: none;
  border-radius: 4px;
  padding-bottom: 0.65625rem;
  padding-left: 16px;
  padding-right: 0;
  padding-top: 0.65625rem;
  color: #32324d;
  font-weight: 400;
  font-size: 0.875rem;
  display: block;
  width: 100%;
  background: inherit;
}

.c8::-webkit-input-placeholder {
  color: #8e8ea9;
  opacity: 1;
}

.c8::-moz-placeholder {
  color: #8e8ea9;
  opacity: 1;
}

.c8:-ms-input-placeholder {
  color: #8e8ea9;
  opacity: 1;
}

.c8::placeholder {
  color: #8e8ea9;
  opacity: 1;
}

.c8[aria-disabled='true'] {
  color: inherit;
}

.c8:focus {
  outline: none;
  box-shadow: none;
}

.c7 {
  border: 1px solid #dcdce4;
  border-radius: 4px;
  background: #ffffff;
  outline: none;
  box-shadow: 0;
  -webkit-transition-property: border-color,box-shadow,fill;
  transition-property: border-color,box-shadow,fill;
  -webkit-transition-duration: 0.2s;
  transition-duration: 0.2s;
}

.c7:focus-within {
  border: 1px solid #4945ff;
  box-shadow: #4945ff 0px 0px 0px 2px;
}

.c1 > * {
  margin-top: 0;
  margin-bottom: 0;
}

.c1 > * + * {
  margin-top: 4px;
}

.c12 path {
  fill: #8e8ea9;
}

.c10 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: 1rem;
  -webkit-align-items: flex-end;
  -webkit-box-align: flex-end;
  -ms-flex-align: flex-end;
  align-items: flex-end;
  -webkit-transform: translateY(-2px);
  -ms-transform: translateY(-2px);
  transform: translateY(-2px);
}

.c10 svg {
  display: block;
  height: 0.25rem;
  -webkit-transform: rotateX(180deg);
  -ms-transform: rotateX(180deg);
  transform: rotateX(180deg);
}

.c13 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: 1rem;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  -webkit-transform: translateY(2px);
  -ms-transform: translateY(2px);
  transform: translateY(2px);
}

.c13 svg {
  display: block;
  height: 0.25rem;
}

<div>
  <div>
    <div
      class="c0 c1"
      spacing="1"
    >
      <label
        class="c2"
        for="number"
        required=""
      >
        <div
          class="c3"
        >
          Default label
          <span
            class="c4 c5"
          >
            *
          </span>
        </div>
      </label>
      <div
        class="c6 c7"
      >
        <input
          aria-disabled="false"
          aria-invalid="false"
          class="c8"
          id="number"
          name="number"
          placeholder="Default placeholder"
          type="text"
          value=""
        />
        <div
          class="c9"
        >
          <button
            aria-hidden="true"
            class="c10"
            data-testid="ArrowUp"
            tabindex="-1"
            type="button"
          >
            <svg
              class="c11 c12"
              fill="none"
              height="1em"
              viewBox="0 0 14 8"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                clip-rule="evenodd"
                d="M14 .889a.86.86 0 01-.26.625L7.615 7.736A.834.834 0 017 8a.834.834 0 01-.615-.264L.26 1.514A.861.861 0 010 .889c0-.24.087-.45.26-.625A.834.834 0 01.875 0h12.25c.237 0 .442.088.615.264a.86.86 0 01.26.625z"
                fill="#32324D"
                fill-rule="evenodd"
              />
            </svg>
          </button>
          <button
            aria-hidden="true"
            class="c13"
            data-testid="ArrowDown"
            tabindex="-1"
            type="button"
          >
            <svg
              class="c11 c12"
              fill="none"
              height="1em"
              viewBox="0 0 14 8"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                clip-rule="evenodd"
                d="M14 .889a.86.86 0 01-.26.625L7.615 7.736A.834.834 0 017 8a.834.834 0 01-.615-.264L.26 1.514A.861.861 0 010 .889c0-.24.087-.45.26-.625A.834.834 0 01.875 0h12.25c.237 0 .442.088.615.264a.86.86 0 01.26.625z"
                fill="#32324D"
                fill-rule="evenodd"
              />
            </svg>
          </button>
        </div>
      </div>
    </div>
  </div>
  <div
    class="c14"
  >
    <p
      aria-live="polite"
      aria-relevant="all"
      id="live-region-log"
      role="log"
    />
    <p
      aria-live="polite"
      aria-relevant="all"
      id="live-region-status"
      role="status"
    />
    <p
      aria-live="assertive"
      aria-relevant="all"
      id="live-region-alert"
      role="alert"
    />
  </div>
</div>
`;
